design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+
+**tl;dr:** *Improve the interactivity of the development story by enabling CLI users to directly
+produce data to Kafka topics through KSQL*
+
+## Motivation and background
+
+KSQL is a powerful tool for exploring Apache Kafka and stream processing paradigms. It allows a 
+developer to easily create Streams and Tables from source topics, apply arbitrary transformations
+and inspect resulting data in Kafka. For developers with relational database experience, the 
+SQL-like language provides a familiar workflow.
+
+After installing KSQL, developers are met with an immediate hurdle - they must have existing data 
+in Kafka that match the data they expect to handle. In a relational database environment, developers
+can use `INSERT INTO ... VALUES` to quickly construct a toy environment for testing. Generating this 
+data KSQL, however, is cumbersome: 
+
+* you require a separate tool (e.g. `kafkacat`, `ksql-datagen`)
+* you must know the deployment topology (e.g. URIs for ZK/Broker)
+* there is no synergy with KSQL (e.g. tools are not *schema aware* and can produce arbitrary data)
+
+This KLIP attempts to integrate producing test data into the KSQL REPL.
+
+## Scope
+
+This KLIP is intended to improve the *development and testing* workflow as opposed to providing
+*production-ready* functionality. The following tasks will be addressed:
+
+* KSQL will be able to create and produce data to arbitrary topics
+* KSQL will be able to validate data when producing to a topic that is registered with a schema
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO
+
+```sql
+INSERT INTO (stream_name|table_name) (column1, column2, ...)
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT INTO` statement can be used to insert new records into a kafka topic (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. The key for the insert statement is taken from the
+column the matches the name of `key` in the `CREATE` statement properties for the corresponding
+source. This functionality will need to be extended when we support structured keys.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (i.e. by following syntax 
+like Hive's `NAMED_STRCUT` construct). 
```

Comment:
Can we include an example of this here, to make sure that the present work can be extended sensibly to support complex types.